OxidosAutomotive/tock
# Main PLL clock implementation for STM32F4xx

## Pull request overview

This pull request adds support for the main PLL clock. Currently, the STM32F4xx
chips use the default HSI (high-speed internal) clock source. Its default
frequency is 16MHz. Most of the time, it may be enough, but some peripherals
require a higher frequency, such as Ethernet. This may be achieved by using the
main PLL clock as the system clock. Additionally, some peripherals clocks rely
on the PLL clock, such as USB OTG FS, the random number generator and SDIO.

# Implemented features

- [x] Default configuration of 96MHz with reduced PLL jitter
- [x] 1MHz frequency precision
- [x] Support for 13-216MHz frequency range
- [x] Support for PLL48CLK output

# Missing features

- [ ] Precision higher than 1MHz
- [ ] Source selection
- [ ] Precise control over the PLL48CLK frequency


## Implementation details

### Why a separate file?

A separate source file has been created for two reasons:

+ **rcc.rs** is already a pretty large file
+ the logic for the PLL clock is kept inside **pll.rs** and the register
operations inside **rcc.rs**

### PLL public interface

The public `Pll` interface resembles `kernel::platform::chip::ClockInterface`,
offering the same methods `enable`, `disable` and `is_enabled`, but returning
`Result` to allow the caller to handle errors. In addition, `get_frequency` and
`set_frequency` offer the possibility to query the frequency of the main PLL
clock and to set it. Furthermore, `get_frequency_pll48` and
`is_pll48_calibrated` allows to query the frequency of the PLL48CLK output and
to check whether it is calibrated or not, i.e. its frequency is exactly 48MHz.

### Frequency output

Due to the hardware structure, it is impossible to configure both outputs
independently. The implementation chose to focus on the main output and try to
configure in the best way the PLL48CLK output: set a 48MHz frequency if
possible, otherwise, the highest value less than 48MHz.

Also, the frequency value is cached to avoid reading the registers every time
it is needed. As a result, the PLL configuration register **must not** be
accessed outside the `Pll` struct. See TODO.

## Testing

This pull requests comes with both unit tests and integration tests. The
documentation explains how to run the tests. The tests were run on the
STM32F429ZI Nucleo-144.

## Documentation

- [x] Internal documentation
- [x] External documentation for public methods.

## Help wanted

+ Couldn't figure how to add static asserts based on the `register_fields!`
macro. They can help to detect pesky bugs when the default values for different
PLL configuration parameters are changed.
+ Run tests on other STM32F4xx based boards.

## TODO

+ Test with an oscilloscope 
+ Add higher precision.
+ Add source switching.
+ Rewrite TRNG peripheral driver to use the new PLL clock struct instead of
accessing the PLLCFGR register directly. See `configure_rng_clock()`.

## Formatting

- [x] Ran `make prepush`
